Goto command support to handle dirty Database state

venkat-iblox opened the pull request:
This PR introduces improvements to the migration workflow by implementing robust handling for dirty states, effective failure management, and cleanup of newer migration files.

Goto command can handle the reversal of dirty DB state if the following args are set:
- `-force-dirty-handling`: Lets the goto module know that the command has to handle the dirty state(required)
- `-cache-dir`: The path where the current migrations(successfully executed) are stored. This is usually a mounted volume to the migration container and persists beyond the migration container's lifecycle. 

###  Key Changes:

1. Handle Dirty State:

- Implements `HandleDirtyState` to apply the last successful migration when a dirty state is detected.

2. Migration Failure Management:

- Integrated `HandleMigrationFailure` to capture and store the last successful migration version in case of failure, ensuring available recovery options.

3. File Cleanup:

- Adds `CleanupFiles` to remove migration files with versions greater than the current migration version after successful execution, maintaining a clean working directory.

Added UTs for the new functionality.

daniel-garcia commented on cmd/migrate/config.go:
```diff
@@ -35,4 +35,8 @@ var (
 	flagConfigFile      = pflag.String("config.file", "", "configuration file name without extension")
+
+	// goto command flags
+	flagDirty   = pflag.Bool("dirty", false, "migration is dirty")
```
maybe this should be a verb instead of a noun. for example, 'handle-dirty', 'force-dirty-handling'. i'm open to the naming here but `dirty` by itself is not invocative about what this does.

venkat-iblox replied:
`force-dirty-handling` seems clear and concise on what it is supposedly doing

daniel-garcia replied:
That works for me

venkat-iblox commented:
@daniel-garcia I am not able to add reviewers to this PR. Please review when you get a chance

venkat-iblox commented on migrate.go:
```diff
@@ -979,3 +1049,162 @@ func (m *Migrate) logErr(err error) {
+	// Determine the last successful migration
+	lastSuccessfulMigration := strconv.Itoa(curVersion)
+	ret := make(chan interface{}, m.PrefetchMigrations)
+	go m.read(curVersion, int(v), ret)
```
@akleinib I am reworking the logic. We don't need this here as the `runMigrations` function can be used to determine last successful migration version

venkat-iblox commented on migrate.go:
```diff
@@ -979,3 +1049,162 @@ func (m *Migrate) logErr(err error) {
+		fileName := file.Name()
+
+		// Check if file is a migration file we want to process
+		if !strings.HasSuffix(fileName, "down.sql") && !strings.HasSuffix(fileName, "up.sql") {
```
I will take a look at this.

venkat-iblox commented on migrate.go:
```diff
@@ -80,6 +87,15 @@ type Migrate struct {
+	ds *dirtyStateHandler
+}
+
+type dirtyStateHandler struct {
```
Ok. I will avoid the stutter and rename

venkat-iblox replied:
@akleinib I have addressed this. is there something I need to fix to resolve this suggestion?

venkat-iblox commented on migrate.go:
```diff
@@ -215,20 +248,57 @@ func (m *Migrate) Migrate(version uint) error {
 	if dirty {
-		return m.unlockErr(ErrDirty{curVersion})
+		if m.ds != nil && m.ds.isDirty {
+			if err = m.unlock(); err != nil {
```
Yes. I am reworking the `HandleDirtyState` func to do so. The current call to `Force` in the `HandleDirtyState` requires the release of lock

venkat-iblox commented on cmd/migrate/config.go:
```diff
@@ -35,4 +35,8 @@ var (
+
+	// goto command flags
+	flagDirty   = pflag.Bool("dirty", false, "migration is dirty")
+	flagPVCPath = pflag.String("intermediate-path", "", "path to the mounted volume which is used to copy the migration files")
```
Sure

akleinib replied:
I still see the `mounted volume` term here.

venkat-iblox commented on migrate.go:
```diff
@@ -979,3 +1049,162 @@ func (m *Migrate) logErr(err error) {
+	return nil
+}
+
+// CopyFiles copies all files from srcDir to destDir.
```
This function along with cleanup shouldn't be public at all. I am fixing this in the current rework

venkat-iblox commented on migrate.go:
```diff
@@ -979,3 +1049,162 @@ func (m *Migrate) logErr(err error) {
+			destFile *os.File
+		)
+		dest := filepath.Join(m.ds.destPath, info.Name())
+		if srcFile, err = os.Open(src); err != nil {
+			return err
+		}
+		defer func(srcFile *os.File) {
+			if err = srcFile.Close(); err != nil {
+				m.logErr(fmt.Errorf("failed to close file %s: %v", destFile.Name(), err))
+			}
+		}(srcFile)
+
+		// Create the destination file
+		if destFile, err = os.Create(dest); err != nil {
+			return err
+		}
+		defer func(destFile *os.File) {
+			if err = destFile.Close(); err != nil {
+				m.logErr(fmt.Errorf("failed to close file %s: %v", destFile.Name(), err))
+			}
+		}(destFile)
+
+		// Copy the file
+		if _, err = io.Copy(destFile, srcFile); err != nil {
+			return err
+		}
+		return os.Chmod(dest, info.Mode())
```
We try this as I rework the logic in the func to use regex to ignore unwanted files

venkat-iblox commented on migrate.go:
```diff
@@ -215,20 +248,57 @@ func (m *Migrate) Migrate(version uint) error {
+		}
+	}
+
+	// Copy migrations to the destination directory,
```
Yes, you are correct. It is currently developed only for file driver

daniel-garcia commented on internal/cli/main.go:
```diff
@@ -29,7 +29,7 @@ const (
            Use -tz option to specify the timezone that will be used when generating non-sequential migrations (defaults: UTC).
 `
-	gotoUsage = `goto V       Migrate to version V`
+	gotoUsage = `goto V [-dirty] [-intermediate-path]     Migrate to version V`
```
`-cache-dir` ?

venkat-iblox commented:
@akleinib @daniel-garcia I have addressed comments except https://github.com/infobloxopen/migrate/pull/36#discussion_r1775534696. To address this comment, we will need a broader work item to introduce new methods in the driver interface and its implementation across various supported drivers

akleinib commented:
> @akleinib @daniel-garcia I have addressed comments except [#36 (comment)](https://github.com/infobloxopen/migrate/pull/36#discussion_r1775534696). To address this comment, we will need a broader work item to introduce new methods in the driver interface and its implementation across various supported drivers

Maybe we can put a limitation that the new options only work if the `file` driver is used for source.

venkat-iblox commented:
> > @akleinib @daniel-garcia I have addressed comments except [#36 (comment)](https://github.com/infobloxopen/migrate/pull/36#discussion_r1775534696). To address this comment, we will need a broader work item to introduce new methods in the driver interface and its implementation across various supported drivers
> 
> Maybe we can put a limitation that the new options only work if the `file` driver is used for source.

Sure

venkat-iblox commented on migrate.go:
```diff
@@ -979,3 +1049,162 @@ func (m *Migrate) logErr(err error) {
+	_, err := os.ReadDir(m.ds.destPath)
+	if err != nil {
+		// If the directory does not exist
+		return err
```
I have addressed this by returning a more meaningful error message.

akleinib commented on migrate_test.go:
```diff
@@ -1414,3 +1417,282 @@ func equalDbSeq(t *testing.T, i int, expected migrationSequence, got *dStub.Stub
+			}
+
+			for _, file := range test.copiedFiles {
+				if _, err := os.Stat(filepath.Join(destDir, file)); os.IsNotExist(err) {
```
Ideally you want to check that `"current.sql"` hasn't been copied.

I don't know if using https://pkg.go.dev/os#ReadDir you can write a helper function returning all the files in a directory you can compare with an expected list of files. It would avoid in the various test-cases the two loops (one checking for existence, one checking for absence).

daniel-garcia replied:
I don't think it would matter. current.sql is a project specific concept, migrate file driver doesn't know anything about things that don't match its pattern.

daniel-garcia commented on migrate.go:
```diff
@@ -979,3 +1083,114 @@ func (m *Migrate) logErr(err error) {
+func (m *Migrate) handleDirtyState() error {
+	// Perform the following actions when the database state is dirty
+	/*
+	   1. Update the source driver to read the migrations from the mounted volume
```
use `cache-dir`

daniel-garcia reviewed (approved):
LGTM. The public interface (cli switches and behaviors) is in decent shape. We can address other concerns in a followup PR
